controllers/cluster_controller.go

```diff
@@ -63,20 +69,74 @@ func (r *ClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 		return ctrl.Result{}, err
 	}
 
-	// Return early if the object or Cluster is paused.
+	// Return if the Cluster is paused.
 	if annotations.IsPaused(cluster, cluster) {
+		r.Log.Info("The cluster is paused.")
 		return ctrl.Result{}, nil
 	}
 
-	// your logic here
+	// Return if the Cluster is deleted.
+	if !cluster.DeletionTimestamp.IsZero() {
+		r.Log.Info("The cluster is deleted.")
+		return ctrl.Result{}, nil
+	}
+
+	// Return if there is no upgrade time scheduled.
+	if getClusterUpgradeTimeAnnotation(cluster) == "" {
+		r.Log.Info("The cluster has no upgrade scheduled.")
+		return ctrl.Result{}, nil
+	}
+
+	// Return if the upgrade release version is not specified.
+	if getClusterUpgradeVersionAnnotation(cluster) == "" {
+		r.Log.Info("The scheduled update at %v can not proceed because no target release version has been set via annotation %v.", getClusterUpgradeTimeAnnotation(cluster), annotation.AWSUpdateScheduleTargetRelease)
+		return ctrl.Result{}, nil
+	}
+
+	return r.ReconcileUpgrade(ctx, cluster)
+}
+
+func (r *ClusterReconciler) ReconcileUpgrade(ctx context.Context, cluster *clusterv1.Cluster) (ctrl.Result, error) {
+	upgradeTime, err := time.Parse(time.RFC822, getClusterUpgradeTimeAnnotation(cluster))
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Return if the scheduled upgrade time is not reached yet.
+	if !upgradeTimeReached(upgradeTime) {
+		r.Log.Info(fmt.Sprintf("The scheduled update time is not reached yet. Cluster will be upgraded at %v.", upgradeTime))
+		return ctrl.Result{}, nil
```

Comment:
if you don't specify reque time in `ctrl.Result` it will never be reconciled again

here I would also try take advantage of the reque and if the upgrade time is very close than reque at exact time of the update

Reply:
That is very interesting. So there is no default time? In that case We should probably always reschedule it.
I'll also include the part for upgrades that happen soon.

Reply:
I don't think so you need to specify it, the benefit you can reconcile all objects different or making calls not too often.
Here's an example how to set it: https://github.com/giantswarm/capa-iam-controller/blob/master/controllers/awsmachinepool_controller.go#L167-L170

Reply:
Thank you for the explanation!
I added my own default reconcile result that will reconcile after 5 minutes. :)